File: Week 2/id_545/LeetCode_94_545.py
## 94.二叉树的中序遍历

# 迭代法
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        if not root: return []
        arr = []
        stack = []
        curr = root
        while curr or stack:
            if curr:
                stack.append(curr)
                curr = curr.left
            else:
                curr = stack.pop()
                arr.append(curr.val)
                curr = curr.right
        return arr

# 递归
# class Solution:
#     def __init__(self):
#         self.arr = []
#     def inorderTraversal(self, root: TreeNode) -> List[int]:
#         if root:
#             self.inorderTraversal(root.left)
#             self.arr.append(root.val)
#             self.inorderTraversal(root.right)
#         return self.arr

File: Week 2/id_545/LeetCode_94_545.js
// 94.二叉树的中序遍历

/*******
// 递归
class Inorder {
    constructor() {
        this.arr = []
    }
    traversal(root) {
        if(root) {
            this.traversal(root.left)
            this.arr.push(root.val)
            this.traversal(root.right)
        }
        return this.arr
    }
}

var inorderTraversal = function(root) {
    let inor = new Inorder()
    return inor.traversal(root)
}
*******/

// 迭代 
var inorderTraversal = function(root) {
    let arr = []
    let stack = []
    let curr = root
    while (curr || stack.length != 0) {
        if (curr) {
            stack.push(curr)
            curr = curr.left
        } else {
            curr = stack.pop()
            arr.push(curr.val)
            curr = curr.right
        }
    }
    return arr
};